lib/apm.js

```diff
@@ -28,4 +32,104 @@ class Instrumentation extends EventEmitter {
   }
 }
 
-module.exports = Instrumentation;
+const apmLoggingEvents = {
+  command: ['commandStarted', 'commandSucceeded', 'commandFailed'],
+  sdam: [
+    'serverOpening',
+    'serverDescriptionChanged',
+    'serverHeartbeatStarted',
+    'serverHeartbeatSucceeded',
+    'serverHeartbeatFailed',
+    'serverClosed',
+    'topologyOpening',
+    'topologyClosed',
+    'topologyDescriptionChanged'
+  ]
+};
+
+class apmReadableStream extends Readable {
+  constructor(client, categories) {
+    super();
+    categories.forEach(category => {
+      attachMonitorListenerCategory(client, category, this);
+    });
+  }
+
+  _read() {
+    return;
+  }
+}
+
+function apmLoggingEventHandler(client) {
+  return function(err) {
+    if (!err) return;
+    if (!client.topology) throw err;
+    client.close(true, () => {
+      throw err;
+    });
+  };
+}
+
+function getMonitoringCategories(categories) {
+  const allCategories = Object.keys(apmLoggingEvents);
+  // enable all monitors if monitor = all
+  if (categories === 'all') return allCategories;
+  // else, ensure `monitors` is an array
+  const categoriesArray = Array.isArray(categories) ? categories : [categories];
+
+  // make sure each category is valid
+  categoriesArray.forEach(c => {
+    if (allCategories.indexOf(c) === -1) {
+      throw new TypeError(`Invalid APM monitoring category specified: ${c}`);
+    }
+  });
+
+  return categoriesArray;
+}
+
+function resolveMonitoringOutputFile(monitorOut) {
+  const defaultFileName = `${process.cwd()}/mongodb-monitoring-${new Date().toISOString()}.log`;
+  if (!monitorOut) return defaultFileName;
+  if (monitorOut === 'stderr' || monitorOut === 'stdout') return monitorOut;
+  return path.resolve(monitorOut);
```

Comment:
What behavior should we expect if someone passes in a unix filepath while on a win32 system?